File: codility/go/8-leader.go
package golang

// https://app.codility.com/programmers/lessons/8-leader/dominator/
func dominator(A []int) int {
	m := make(map[int]int, len(A))
	for i, a := range A {
		m[a]++
		if m[a] > len(A)/2 {
			return i
		}
	}

	return -1
}

// https://app.codility.com/programmers/lessons/8-leader/equi_leader/
// insight:
// for a leader to apply to both sides of the slice, it also has to be a leader in the entire, combined slice
// find it that way first, then move the index and count the indexes the leader applies for
func equiLeader(A []int) int {
	if len(A) < 2 {
		return 0
	}

	var leader int
	m := make(map[int]int, len(A))
	for _, a := range A {
		m[a]++
		if m[a] > len(A)/2 {
			leader = a
		}
	}
	// this check is here to confirm that leader is not zero only because of integer initialization, but true value found in A
	if _, ok := m[leader]; !ok {
		return 0
	}

	leftCount := 0
	rightCount := m[leader]
	leaderIndexes := 0
	// We don't need to check the last idx because there is never an equvalent leader on the right side (length zero)
	for i := 0; i < len(A)-1; i++ {
		leftLen := i + 1
		rightLen := len(A) - i - 1

		// if current idx is a leader, update counts
		if A[i] == leader {
			leftCount++
			rightCount--
		}

		// add to count if leader still applies both left and right
		if leftCount > leftLen/2 && rightCount > rightLen/2 {
			leaderIndexes++
		}
	}

	return leaderIndexes
}

File: codility/py/8-leader.py
def dominator(A):
    d = {}
    for i, el in enumerate(A):
        if el in d:
            d[el] += 1
        else:
            d[el] = 1

        if d[el] > len(A) // 2:
            return i

    return -1

assert dominator([3, 4, 3, 2, 3, -1, 3, 3]) == 7
assert dominator([1, 2, 3, 4, 5]) == -1
assert dominator([1, 2, 3, 4, 5, 5, 5]) == -1
assert dominator([1, 2, 3, 4, 5, 5, 5, 5,5]) == 8

def equi_leader(A):
    # Find dominator - equileader will be the same value
    N = len(A)
    freq = {}
    for n in A:
        freq[n] = freq.get(n,0)+1
    dominator, dominator_count = None,0
    for k,v in freq.items():
        if v > N//2:
            dominator = k
            dominator_count = v

    if dominator is None:
        return 0

    # Count equi leaders. Start from left
    left_count, result = 0,0
    for i in range(N-1):
        n = A[i]
        if n == dominator:
            left_count += 1

        left_len = i+1
        right_len = N - 1 - i
        right_count = dominator_count - left_count

        # Verfify leader dominates both sides
        if left_count > left_len//2 and right_count>right_len//2:
            result +=1

    return result